packages/SwingSet/src/vats/comms/state.js

```diff
@@ -128,6 +131,98 @@ export function makeState(syscall, identifierBase = 0) {
     }
   }
 
+  function deleteLocalPromiseState(lpid) {
+    store.delete(`${lpid}.status`);
+    store.delete(`${lpid}.decider`);
+    store.delete(`${lpid}.kernelSubscribed`);
+    store.delete(`${lpid}.subscribers`);
+    store.delete(`${lpid}.data.body`);
+    store.delete(`${lpid}.data.slots`);
+    store.delete(`${lpid}.refCount`);
+  }
+
+  function deleteLocalPromise(lpid) {
+    const status = store.getRequired(`${lpid}.status`);
+    switch (status) {
+      case 'unresolved':
+        break;
+      case 'fulfilled':
+        break;
+      case 'rejected':
+        break;
+      default:
+        assert.fail(X`unknown status for ${lpid}: ${status}`);
+    }
+    deleteLocalPromiseState(lpid);
+  }
+
+  const deadLocalPromises = new Set();
+
+  /**
+   * Increment the reference count associated with some local object.
+   *
+   * Note that currently we are only reference counting promises, but ultimately
+   * we intend to keep track of all local objects.
+   *
+   * @param {string} lref  Ref of the local object whose refcount is to be incremented.
+   * @param {string} _tag  Descriptive label for use in diagnostics
+   */
+  function incrementRefCount(lref, _tag) {
+    if (lref && parseLocalSlot(lref).type === 'promise') {
+      const refCount = Number(store.get(`${lref}.refCount`)) + 1;
+      // cdebug(`++ ${lref}  ${tag} ${refCount}`);
+      store.set(`${lref}.refCount`, `${refCount}`);
+    }
+  }
+
+  /**
+   * Decrement the reference count associated with some local object.
+   *
+   * Note that currently we are only reference counting promises, but ultimately
+   * we intend to keep track of all local objects.
+   *
+   * @param {string} lref  Ref of the local object whose refcount is to be decremented.
+   * @param {string} tag  Descriptive label for use in diagnostics
+   * @returns {boolean} true if the reference count has been decremented to
+   *    zero, false if it is still non-zero
+   * @throws if this tries to decrement the reference count below zero.
+   */
+  function decrementRefCount(lref, tag) {
+    if (lref && parseLocalSlot(lref).type === 'promise') {
+      let refCount = Number(store.get(`${lref}.refCount`));
+      assert(refCount > 0n, X`refCount underflow {lref} ${tag}`);
+      refCount -= 1;
+      // cdebug(`-- ${lref}  ${tag} ${refCount}`);
+      store.set(`${lref}.refCount`, `${refCount}`);
+      if (refCount === 0) {
+        deadLocalPromises.add(lref);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  function purgeDeadLocalPromises() {
+    if (enableLocalPromiseGC) {
+      for (const lpid of deadLocalPromises.values()) {
+        const refCount = Number(store.get(`${lpid}.refCount`));
+        if (refCount === 0) {
```

Comment:
Can refCount ever be nonzero? I use this sort of check a lot, but now I'm trying to think about whether it's actually possible to resurrect an lref once its refCount drops to zero. I don't think so.

If that were possible, then this loop might not handle a case where the refcount on X dropped to zero (putting the lref into `deadLocalPromises`), some "Y" thing brought it back up to 1, something drops the Y refcount to 0, then `purgeDeadLocalPromises` is invoked which starts a loop, it sees the X refcount as nonzero so it keeps moving on, it sees Y refcount as zero so it deletes it, which reduces the X refcount to zero, so `decrementRefCount` does a new `deadLocalPromises.add(X)`. But from my experiments, adding X to a set while it's iterating *and has iterated past X* doesn't cause X to be checked a second time. So that second "`decrementRefCount` to zero" call wouldn't trigger the proper deletion.

But I don't think we can get into this situation anyways, because once an lref has a refcount of zero, there shouldn't be any way to bring it back up again.

This might call for the refCount check to go away, or maybe it should just be an assertion.

Reply:
At this point I don't think the refCount can ever be non-zero.  This was how it was written in the kernel, which may have slightly more complex behavior in this regard, though I suspect it's just suspenders-and-belt.  I think making it an assertion is almost certainly the right thing, as having it be non-zero at this point would lead to the kind of whackiness as you describe.  I will make that change.